src/dune_rules/dune_load.ml

```diff
@@ -195,28 +195,54 @@ let interpret ~dir ~project ~(dune_file : File_tree.Dune_file.t) =
   | Plain -> Literal (Dune_file.parse static ~dir ~file ~project)
 
 let load ~ancestor_vcs =
+  let open Fiber.O in
   File_tree.init ~ancestor_vcs ~recognize_jbuilder_projects:false;
-  let _, vcs, projects =
-    let f dir (ancestor_vcs, vcs, projects) =
-      let vcs =
-        match File_tree.Dir.vcs dir with
-        | Some repository -> Path.Map.set vcs repository.root repository
-        | None -> vcs
+  let+ vcs, projects, dune_files =
+    Memo.Build.run
+      (let module Outcome = Monoid.Make (struct
+         type t =
+           Vcs.t Path.Map.t
+           * Dune_project.t Appendable_list.t
+           * (Path.Source.t * Dune_project.t * File_tree.Dune_file.t)
+             Appendable_list.t
+
+         let empty =
+           (Path.Map.empty, Appendable_list.empty, Appendable_list.empty)
+
+         let combine (a1, a2, a3) (b1, b2, b3) =
```

Comment:
Maybe we should add `Product3` to `Monoid` to avoid writing this boilerplate here :) We already have a 2-way `Product` there.

Reply:
(More generally, it would be nice to have a ppx for deriving monoids.)

Reply:
Sounds good. I did that. For the ppx, it's a bit of a slippery slope. Ppx rewriters are a bit more difficult to develop, maintain and learn that plain OCaml libraries. For things such as `[@@deriving]`, the win is clear, for other things it can lead to a code base that is more obscure than necessary.

Reply:
Yes, I meant `[@@deriving monoid]`  which would be particularly convenient for deriving monoids for record types, but I agree that it's best to keep things simple here in Dune.